packages/lsp/src/adapters/adapter.ts

```diff
@@ -718,4 +733,41 @@ export abstract class WidgetLSPAdapter<
     this._updateFinished = promise.catch(console.warn);
     await this.updateFinished;
   }
+
+  private _shouldUpdateVirtualDocument(languages: string[]): boolean {
+    const { languageServerManager } = this.connectionManager;
+    let matchedServer = false;
+    for (const language of languages) {
+      if (languageServerManager.getMatchingServers({ language })) {
+        matchedServer = true;
+        break;
+      }
+    }
+    return (
+      languageServerManager.isEnabled &&
+      matchedServer &&
+      this.options.featureManager.features.length > 0
+    );
+  }
+
+  private _onLspSessionOrFeatureChanged(): void {
+    if (!this._virtualDocument) {
+      return;
+    }
+
+    const languages = [
+      this._virtualDocument.language,
+      ...[...this._virtualDocument.foreignDocuments.values()].map(
+        it => it.language
+      )
+    ];
```

Comment:
I am afraid that this is a chicken and egg problem. Downstream we do not know all the languages until we updated the document as this is when the extraction happens.